### PR TITLE
Finish IO implementation 

### PR DESCRIPTION
**Pull request by LordRaydenMK**

### CODE REVIEW

File: src/main/kotlin/io/arrowkt/service/interpreter/AccountService.kt [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+        Kleisli { repo ->
+            EitherT(
+                repo.query(no).flatMap { accountOptionOrError ->
+                    accountOptionOrError.fold(
+                        { IO { MiscellaneousDomainExceptions(it).left() } },
+                        { accountOption ->
+                            accountOption.fold(
+                                { IO { NonExistingAccount(no).left() } },
+                                {
+                                    val updated = when (dc) {
```
Comment (LordRaydenMK) [inline src/main/kotlin/io/arrowkt/service/interpreter/AccountService.kt:110]: without the temporary val type inference fails....

File: src/main/kotlin/io/arrowkt/model/Account.kt [hunk 2]
```diff
@@ -38,7 +35,10 @@ sealed class Account(
             cd.map { c ->
                 if (c.isBefore(od)) "Close date $c cannot be earlier than open date $od".invalidNel()
                 else Pair(od.some(), cd).validNel()
-            }.getOrElse { Pair(od.some(), cd).validNel() }
```
Comment (LordRaydenMK) [inline src/main/kotlin/io/arrowkt/model/Account.kt:41]: This breaks kotlin... so I copied the implementation

Comment (pakoito) [reply src/main/kotlin/io/arrowkt/model/Account.kt:41]: How? Inference?

File: src/main/kotlin/io/arrowkt/repository/interpreter/AccountRepository.kt [hunk 3]
```diff
@@ -11,7 +11,7 @@ import io.arrowkt.repository.AccountRepository
 import io.arrowkt.today
 import java.time.LocalDate
 
-object AccountInMemoryRepository : AccountRepository {
+class AccountInMemoryRepository : AccountRepository {
```
Comment (pakoito) [inline src/main/kotlin/io/arrowkt/repository/interpreter/AccountRepository.kt:14]: Why the change?

Comment (pakoito) [reply src/main/kotlin/io/arrowkt/repository/interpreter/AccountRepository.kt:14]: Oh, right, state :D

File: src/main/kotlin/io/arrowkt/service/interpreter/AccountService.kt [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+object AccountService : AccountService<Account, Amount, Balance> {
+
+    override fun open(
+        no: String,
+        name: String,
+        rate: Option<BigDecimal>,
+        openingDate: Option<LocalDate>,
+        accountType: AccountType
+    ): AccountOperation<Account> = Kleisli.invoke { repo ->
+        EitherT(
```
Comment (pakoito) [inline src/main/kotlin/io/arrowkt/service/interpreter/AccountService.kt:27]: The helper function, please :D Else it'll be too hard to read.

**Review by pakoito — APPROVED** [review]
LGTM! A couple of questions

File: src/main/kotlin/io/arrowkt/repository/interpreter/AccountRepository.kt (hunk 3 above)
Comment (LordRaydenMK) [reply src/main/kotlin/io/arrowkt/repository/interpreter/AccountRepository.kt:14]: yup, each use case in the main app creates a new repository

File: src/main/kotlin/io/arrowkt/model/Account.kt (hunk 2 above)
Comment (LordRaydenMK) [reply src/main/kotlin/io/arrowkt/model/Account.kt:41]: Not really... more like the compiler screaming in pain

```
Error:Kotlin: [Internal Error] java.lang.IllegalStateException: Backend Internal error: Exception during code generation
Cause: Back-end (JVM) Internal error: Cannot serialize error type: [ERROR : Unknown type parameter 0. Please try recompiling module containing "Class 'arrow.data.ValidatedKt'"]
Cause: Cannot serialize error type: [ERROR : Unknown type parameter 0. Please try recompiling module containing "Class 'arrow.data.ValidatedKt'"]
File being compiled at position: (38,25) in /home/coka/IdeaProjects/frdomain.kt/src/main/kotlin/io/arrowkt/model/Account.kt
The root cause was thrown at: SerializerExtension.kt:83
File being compiled at position: file:///home/coka/IdeaProjects/frdomain.kt/src/main/kotlin/io/arrowkt/model/Account.kt
The root cause was thrown at: ExpressionCodegen.java:322
	at org.jetbrains.kotlin.codegen.CompilationErrorHandler.lambda$static$0(CompilationErrorHandler.java:24)
	at org.jetbrains.kotlin.codegen.PackageCodegenImpl.generate(PackageCodegenImpl.java:75)
	at org.jetbrains.kotlin.codegen.DefaultCodegenFactory.generatePackage(CodegenFactory.kt:99)
	at org.jetbrains.kotlin.codegen.DefaultCodegenFactory.generateModule(CodegenFactory.kt:70)
	at org.jetbrains.kotlin.codegen.KotlinCodegenFacade.doGenerateFiles(KotlinCodegenFacade.java:47)
	at org.jetbrains.kotlin.codegen.KotlinCodegenFacade.compileCorrectFiles(KotlinCodegenFacade.java:39)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinToJVMBytecodeCompiler.generate(KotlinToJVMBytecodeCompiler.kt:455)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinToJVMBytecodeCompiler.compileModules$cli(KotlinToJVMBytecodeCompiler.kt:154)
	at org.jetbrains.kotlin.cli.jvm.K2JVMCompiler.doExecute(K2JVMCompiler.kt:131)
	at org.jetbrains.kotlin.cli.jvm.K2JVMCompiler.doExecute(K2JVMCompiler.kt:54)
	at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.kt:84)
	at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.kt:42)
	at org.jetbrains.kotlin.cli.common.CLITool.exec(CLITool.kt:103)
	at org.jetbrains.kotlin.daemon.CompileServiceImpl$compile$$inlined$ifAlive$lambda$1.invoke(CompileServiceImpl.kt:436)
	at org.jetbrains.kotlin.daemon.CompileServiceImpl$compile$$inlined$ifAlive$lambda$1.invoke(CompileServiceImpl.kt:102)
	at org.jetbrains.kotlin.daemon.CompileServiceImpl$doCompile$$inlined$ifAlive$lambda$2.invoke(CompileServiceImpl.kt:1023)
	at org.jetbrains.kotlin.daemon.CompileServiceImpl$doCompile$$inlined$ifAlive$lambda$2.invoke(CompileServiceImpl.kt:102)
	at org.jetbrains.kotlin.daemon.common.DummyProfiler.withMeasure(PerfUtils.kt:137)
	at org.jetbrains.kotlin.daemon.CompileServiceImpl.checkedCompile(CompileServiceImpl.kt:1065)
	at org.jetbrains.kotlin.daemon.CompileServiceImpl.doCompile(CompileServiceImpl.kt:1022)
	at org.jetbrains.kotlin.daemon.CompileServiceImpl.compile(CompileServiceImpl.kt:434)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at sun.rmi.server.UnicastServerRef.dispatch(UnicastServerRef.java:324)
	at sun.rmi.transport.Transport$1.run(Transport.java:200)
	at sun.rmi.transport.Transport$1.run(Transport.java:197)
	at java.security.AccessController.doPrivileged(Native Method)
	at sun.rmi.transport.Transport.serviceCall(Transport.java:196)
	at sun.rmi.transport.tcp.TCPTransport.handleMessages(TCPTransport.java:568)
	at sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run0(TCPTransport.java:826)
	at sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.lambda$run$0(TCPTransport.java:683)
	at java.security.AccessController.doPrivileged(Native Method)
	at sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run(TCPTransport.java:682)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: org.jetbrains.kotlin.codegen.CompilationException: Back-end (JVM) Internal error: Cannot serialize error type: [ERROR : Unknown type parameter 0. Please try recompiling module containing "Class 'arrow.data.ValidatedKt'"]
Cause: Cannot serialize error type: [ERROR : Unknown type parameter 0. Please try recompiling module containing "Class 'arrow.data.ValidatedKt'"]
File being compiled at position: (38,25) in /home/coka/IdeaProjects/frdomain.kt/src/main/kotlin/io/arrowkt/model/Account.kt
The root cause was thrown at: SerializerExtension.kt:83
	at org.jetbrains.kotlin.codegen.ExpressionCodegen.genQualified(ExpressionCodegen.java:322)
	at org.jetbrains.kotlin.codegen.ExpressionCodegen.genQualified(ExpressionCodegen.java:281)
	at org.jetbrains.kotlin.codegen.ExpressionCodegen.gen(ExpressionCodegen.java:354)
	at org.jetbrains.kotlin.codegen.CallGenerator$DefaultCallGenerator.genValueAndPut(CallGenerator.kt:70)
	at org.jetbrains.kotlin.codegen.CallBasedArgumentGenerator.generateExpression(CallBasedArgumentGenerator.kt:42)
	at org.jetbrains.kotlin.codegen.ArgumentGenerator.generate(ArgumentGenerator.kt:68)
	at org.jetbrains.kotlin.codegen.ExpressionCodegen.invokeMethodWithArguments(ExpressionCodegen.java:2490)
	at org.jetbrains.kotlin.codegen.ExpressionCodegen.invokeMethodWithArguments(ExpressionCodegen.java:2462)
	at org.jetbrains.kotlin.codegen.Callable$invokeMethodWithArguments$1.invoke(Callable.kt:41)
	at org.jetbrains.kotlin.codegen.Callable$invokeMethodWithArguments$1.invoke(Callable.kt:13)
	at org.jetbrains.kotlin.codegen.OperationStackValue.putSelector(StackValue.kt:79)
	at org.jetbrains.kotlin.codegen.StackValue.put(StackValue.java:118)
	at org.jetbrains.kotlin.codegen.StackValue.put(StackValue.java:107)
	at org.jetbrains.kotlin.codegen.ExpressionCodegen.putStackValue(ExpressionCodegen.java:378)
	at org.jetbrains.kotlin.codegen.ExpressionCodegen.gen(ExpressionCodegen.java:363)
	at org.jetbrains.kotlin.codegen.ExpressionCodegen.returnExpression(ExpressionCodegen.java:1687)
	at org.jetbrains.kotlin.codegen.FunctionGenerationStrategy$FunctionDefault.doGenerateBody(FunctionGenerationStrategy.java:64)
	at org.jetbrains.kotlin.codegen.FunctionGenerationStrategy$CodegenBased.generateBody(FunctionGenerationStrategy.java:84)
	at org.jetbrains.kotlin.codegen.FunctionCodegen.generateMethodBody(FunctionCodegen.java:666)
	at org.jetbrains.kotlin.codegen.FunctionCodegen.generateMethodBody(FunctionCodegen.java:473)
	at org.jetbrains.kotlin.codegen.FunctionCodegen.generateMethod(FunctionCodegen.java:259)
	at org.jetbrains.kotlin.codegen.FunctionCodegen.generateMethod(FunctionCodegen.java:175)
	at org.jetbrains.kotlin.codegen.FunctionCodegen.gen(FunctionCodegen.java:146)
	at org.jetbrains.kotlin.codegen.MemberCodegen.genSimpleMember(MemberCodegen.java:197)
	at org.jetbrains.kotlin.codegen.ClassBodyCodegen.generateDeclaration(ClassBodyCodegen.java:166)
	at org.jetbrains.kotlin.codegen.ClassBodyCodegen.generateBody(ClassBodyCodegen.java:86)
	at org.jetbrains.kotlin.codegen.MemberCodegen.generate(MemberCodegen.java:129)
	at org.jetbrains.kotlin.codegen.MemberCodegen.genClassOrObject(MemberCodegen.java:302)
	at org.jetbrains.kotlin.codegen.MemberCodegen.genClassOrObject(MemberCodegen.java:286)
	at org.jetbrains.kotlin.codegen.MemberCodegen.genClassOrObject(MemberCodegen.java:312)
	at org.jetbrains.kotlin.codegen.ClassBodyCodegen.generateBody(ClassBodyCodegen.java:102)
	at org.jetbrains.kotlin.codegen.MemberCodegen.generate(MemberCodegen.java:129)
	at org.jetbrains.kotlin.codegen.MemberCodegen.genClassOrObject(MemberCodegen.java:302)
	at org.jetbrains.kotlin.codegen.MemberCodegen.genClassOrObject(MemberCodegen.java:286)
	at org.jetbrains.kotlin.codegen.PackageCodegenImpl.generateClassOrObject(PackageCodegenImpl.java:171)
	at org.jetbrains.kotlin.codegen.PackageCodegenImpl.generateClassesAndObjectsInFile(PackageCodegenImpl.java:87)
	at org.jetbrains.kotlin.codegen.PackageCodegenImpl.generateFile(PackageCodegenImpl.java:130)
	at org.jetbrains.kotlin.codegen.PackageCodegenImpl.generate(PackageCodegenImpl.java:67)
	... 36 more
Caused by: java.lang.IllegalStateException: Cannot serialize error type: [ERROR : Unknown type parameter 0. Please try recompiling module containing "Class 'arrow.data.ValidatedKt'"]
	at org.jetbrains.kotlin.serialization.SerializerExtension.serializeErrorType(SerializerExtension.kt:83)
	at org.jetbrains.kotlin.codegen.serialization.JvmSerializerExtension.serializeErrorType(JvmSerializerExtension.kt:232)
	at org.jetbrains.kotlin.serialization.DescriptorSerializer.type$serialization(DescriptorSerializer.kt:536)
	at org.jetbrains.kotlin.serialization.DescriptorSerializer.typeArgument(DescriptorSerializer.kt:639)
	at org.jetbrains.kotlin.serialization.DescriptorSerializer.fillFromPossiblyInnerType(DescriptorSerializer.kt:607)
	at org.jetbrains.kotlin.serialization.DescriptorSerializer.type$serialization(DescriptorSerializer.kt:565)
	at org.jetbrains.kotlin.serialization.DescriptorSerializer.type$serialization(DescriptorSerializer.kt:589)
	at org.jetbrains.kotlin.serialization.DescriptorSerializer.typeArgument(DescriptorSerializer.kt:639)
	at org.jetbrains.kotlin.serialization.DescriptorSerializer.fillFromPossiblyInnerType(DescriptorSerializer.kt:607)
	at org.jetbrains.kotlin.serialization.DescriptorSerializer.type$serialization(DescriptorSerializer.kt:565)
	at org.jetbrains.kotlin.serialization.DescriptorSerializer.functionProto(DescriptorSerializer.kt:315)
	at org.jetbrains.kotlin.codegen.ClosureCodegen.generateKotlinMetadataAnnotation(ClosureCodegen.java:249)
	at org.jetbrains.kotlin.codegen.MemberCodegen.generate(MemberCodegen.java:136)
	at org.jetbrains.kotlin.codegen.ExpressionCodegen.genClosure(ExpressionCodegen.java:1025)
	at org.jetbrains.kotlin.codegen.ExpressionCodegen.genClosure(ExpressionCodegen.java:995)
	at org.jetbrains.kotlin.codegen.ExpressionCodegen.visitLambdaExpression(ExpressionCodegen.java:986)
	at org.jetbrains.kotlin.codegen.ExpressionCodegen.visitLambdaExpression(ExpressionCodegen.java:111)
	at org.jetbrains.kotlin.psi.KtLambdaExpression.accept(KtLambdaExpression.java:39)
	at org.jetbrains.kotlin.codegen.ExpressionCodegen.genQualified(ExpressionCodegen.java:299)
	... 73 more
```

File: src/main/kotlin/io/arrowkt/service/interpreter/AccountService.kt (hunk 4 above)
Comment (LordRaydenMK) [reply src/main/kotlin/io/arrowkt/service/interpreter/AccountService.kt:27]: the Tagless version is easier on the eye :wink: 

Type inference failed me for the helper function.... I couldn't figure it out...

File: src/main/kotlin/io/arrowkt/model/Account.kt (hunk 2 above)
Comment (pakoito) [reply src/main/kotlin/io/arrowkt/model/Account.kt:41]: yikes

**Comment by pakoito** [comment]
Let's merge then
